### ISSUE TITLE
cmd/gofmt: non-idempotent formatting of line comments

### ISSUE THREAD

**Issue by cezarsa** [golang/go#13702]
tl;dr: Access http://play.golang.org/p/Ry5OUCK43C; Click Format; Click Format.

After running the following snippet through `gofmt -w`:

``` go
package main

//line x:1


func main() {
}
```

The output is:

``` go
package main

//line x:1

func main() {
}
```

This output looks correct with a single newline between `//line` and `func main`. However if we run `gofmt -w` again on the same file the output is different:

``` go
package main

//line x:1
func main() {
}
```

Now the newline between `//line` and `func main` was removed. This behavior is different from what would happen if instead of `//line` we had a regular commented line.

This was tested both on master and 1.5.1:

```
$ go version
go version devel +97f854c Sat Dec 19 10:00:04 2015 +0000 darwin/amd64
```

**Comment by rsc** [comment]
Unfortunate but probably shouldn't be running gofmt on files with //line comments anyway. Can wait for Go 1.7.

**Comment by bradfitz** [comment]
@griesemer, do you still want to do this for Go 1.9? It missed the past 4 releases and still repros.

**Comment by griesemer** [comment]
@bradfitz I will need to see who complicated the fix is, first. I won't get to this before June 1.

**Comment by griesemer** [comment]
I don't think should do anything here: Fixing this would mean that once we see a //line directive we're not allowed to move anything that follows later to a different line. At the very least, that is the assumption we have to make until we see the next //line comment. But that interferes directly with gofmt's main purpose which is to organize white space for uniform formatting. One "solution" would be to not move any lines that follow a //line comment. But it's pretty tricky to get this correct under all circumstances. I'm not convinced the amount of complexity is worth the few cases where it might matter. I don't want to go down that route.

The situation is additionally complicated by the fact that the [go/printer](https://golang.org/pkg/go/printer/) has a [mode](https://golang.org/pkg/go/printer/#Mode) that emits //line comments.

I've recently fixed https://github.com/golang/go/issues/5945 which makes sure that the above-mentioned mode only prints //line directives immediately before non-whitespace. Code that is printed that way will pass through gofmt undisturbed.

In other words, don't use //line directives in front of empty lines - use them where it matters. Or don't gofmt such files.

Closing as unfortunate.
